tests/48admin.pl

```diff
@@ -36,3 +38,188 @@
+
+            matrix_send_room_text_message_synced( $user, $room_id,
+               body => "Message $msgnum",
+            )
+         }, foreach => [ 1 .. 10 ])
+      })->then( sub {
+         ( $last_event_id ) = @_;
+
+         do_request_json_for( $user,
+            method  => "POST",
+            uri     => "/r0/admin/purge_history/$room_id/$last_event_id",
+            content => {}
+         )->main::expect_http_403;  # Must be server admin
+      })->then( sub {
+         matrix_sync( $user )
+      })->then( sub {
+         do_request_json_for( $admin,
+            method  => "POST",
+            uri     => "/r0/admin/purge_history/$room_id/$last_event_id",
+            content => {}
+         )
+      })->then( sub {
+         matrix_sync_again( $user )
+      })->then( sub {
+         matrix_sync( $user )
+      })->then( sub {
+         my ( $body ) = @_;
+
+         assert_json_keys( $body->{rooms}{join}, $room_id );
+         my $room =  $body->{rooms}{join}{$room_id};
+
+         log_if_fail( "Room", $room->{timeline}{events} );
+
+         # The only message event should be the last one.
+         all {
+            $_->{type} ne "m.room.message" || $_->{event_id} eq $last_event_id
+         } @{ $room->{timeline}{events} } or die "Expected no message events";
+
+         # Ensure we still see the state.
+         foreach my $expected_type( qw(
+            m.room.create
+            m.room.member
+            m.room.power_levels
+            m.room.name
+         ) ) {
+            any { $_->{type} eq $expected_type } @{ $room->{state}{events} }
+               or die "Expected state event of type $expected_type";
+         }
+
+         Future->done( 1 );
+      })
+   };
+
+test "Can backfill purged history",
+   requires => [ local_admin_fixture(), local_user_and_room_fixtures(),
+                 remote_user_fixture(), qw( can_paginate_room_remotely ) ],
+
+   do => sub {
+      my ( $admin, $user, $room_id, $remote_user ) = @_;
+
+      my @event_ids;
+      my $last_event_id;
+
+      matrix_invite_user_to_room( $user, $remote_user, $room_id )
+      ->then( sub {
+         matrix_join_room( $remote_user, $room_id )
+      })->then( sub {
+         matrix_put_room_state( $user, $room_id,
+            type    => "m.room.name",
+            content => { name => "A room name" },
+         )
+      })->then( sub {
+         matrix_sync( $remote_user )
+      })->then( sub {
+         repeat( sub {
+            my $msgnum = $_[0];
+
+            matrix_send_room_text_message_synced( $user, $room_id,
+               body => "Message $msgnum",
+            )->on_done( sub { push @event_ids, @_[0]; } )
+         }, foreach => [ 0 .. 4 ])
+      })->then( sub {
+         repeat( sub {
+            my $msgnum = $_[0];
+
+            matrix_send_room_text_message_synced( $remote_user, $room_id,
+               body => "Message $msgnum",
+            )->on_done( sub { push @event_ids, @_[0]; } )
+         }, foreach => [ 5 .. 9 ])
```

Comment:
> Why are half the messages sent by the local user and half by the remote?

To ensure that both local and remote messages are handled correctly by the purge and backfill.

> Not really. There's a helper function for inserting filler messages in https://github.com/matrix-org/sytest/blob/develop/tests/10apidoc/90synced.pl

I'd need to do something custom since I've got an `on_done`. I'm not sure its worth the complexity to try and do this.